syntax = "proto3";

service Master {
  rpc InvokeMapper (MapperParameters) returns (MapperResponse) {}
  rpc InvokeReducer (ReducerParameters) returns (ReducerResponse) {}
}
message Centroid {
  int32 id = 1;
  float x = 2;
  float y = 3;
}
message MapperParameters {
  int32 n_reducers = 1;
  int32 range_start = 2;
  int32 range_end = 3;
  repeated Centroid centroids = 4;
}
message MapperResponse {
  int32 id = 1;
  string status = 2;
}
message ReducerParameters {
  int32 n_mappers = 1;
  repeated Centroid centroids = 2;
}
message ReducerOutputEntry {
  int32 current_centroid_id = 1;
  float updated_centroid_x = 2;
  float updated_centroid_y = 3;
}
message ReducerResponse {
  int32 id = 1;
  string status = 2;
  repeated ReducerOutputEntry entries = 3;
}

service Reducer {
  rpc RequestInput (ReducerInputRequest) returns (ReducerInputResponse) {}
}
message ReducerInputRequest {
  int32 reducer_id = 1;
}
message PartitionEntry {
  int32 key = 1;
  int32 point_id = 2;
  float x = 3;
  float y = 4;
}
message ReducerInputResponse {
  int32 mapper_id = 1;
  string status = 2;
  repeated PartitionEntry partition = 3;
}